src/main/java/com/scalar/db/storage/cosmos/MutateStatementHandler.java

```diff
@@ -107,10 +144,43 @@ protected String makeConditionalQuery(Mutation mutation) {
     return builder.getQuery();
   }
 
+  protected void throwException(CosmosException exception) throws ExecutionException {
+    LOGGER.error(exception.getMessage());
+    int statusCode = exception.getSubStatusCode();
+
+    if (statusCode == CosmosErrorCode.PRECONDITION_FAILED.get()) {
+      throw new NoMutationException("no mutation was applied.");
+    } else if (statusCode == CosmosErrorCode.RETRY_WITH.get()) {
+      throw new RetriableExecutionException(exception.getMessage(), exception);
+    }
+
+    throw new ExecutionException(exception.getMessage(), exception);
+  }
+
   private Map<String, Object> toMap(Collection<Value> values) {
     MapVisitor visitor = new MapVisitor();
     values.forEach(v -> v.accept(visitor));
 
     return visitor.get();
   }
+
+  @VisibleForTesting
+  Record makeDummyRecord() {
```

Comment:
```suggestion
  Record makeEmptyRecord() {
```
Dummy sounds like it is used only for testing but it is actually used for delete, so `empty` sounds better I feel.

Reply:
The method was removed.